src/backtrace.rs

```diff
@@ -0,0 +1,105 @@
+// SPDX-FileCopyrightText: 2023 Rivos Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+/// Get an array of backtrace addresses.
+///
+/// This needs `force-frame-pointers` enabled.
+use alloc::fmt::{Display, Formatter, Result};
+use core::arch::asm;
+
+extern "C" {
+    static _stack_start: u8;
+    static _stack_end: u8;
+}
+
+#[derive(Copy, Clone)]
+pub enum BTReturnAddress {
+    ReturnAddress(u64),
+    InvalidFramePointer(u64),
+    InvalidReturnAddress(u64),
+}
+
+impl Display for BTReturnAddress {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            BTReturnAddress::ReturnAddress(addr) => {
+                writeln!(f, "0x{addr:x}")
+            }
+            BTReturnAddress::InvalidFramePointer(addr) => {
+                writeln!(f, "Invalid frame pointer: 0x{addr:x}")
+            }
+            BTReturnAddress::InvalidReturnAddress(addr) => {
+                writeln!(f, "Null address at frame pointer: 0x{addr:x}")
+            }
+        }
+    }
+}
+
+pub struct BackTrace {
+    fp: Option<u64>,
+    stack_start: u64,
+    stack_end: u64,
+    stack: &'static [u64],
+}
+
+impl BackTrace {
+    fn new(fp: u64) -> Self {
+        // Safe because addresses from linker
+        let stack_start = unsafe { core::ptr::addr_of!(_stack_start) as u64 };
+        let stack_end = unsafe { core::ptr::addr_of!(_stack_end) as u64 };
+        let stack = unsafe {
+            core::slice::from_raw_parts(
+                stack_start as *const u64,
+                (stack_end - stack_start) as usize,
+            )
+        };
+
+        Self {
+            fp: Some(fp),
+            stack_start,
+            stack_end,
+            stack,
+        }
+    }
+}
+
+impl Iterator for BackTrace {
+    type Item = BTReturnAddress;
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(mut fp) = self.fp {
+            if fp < self.stack_start + 2 || fp > self.stack_end {
+                self.fp = None;
+                return Some(BTReturnAddress::InvalidFramePointer(fp));
+            }
+
+            let offset = ((fp - self.stack_start - 2) / 8) as usize;
+            if offset > 0 && let Some(address) = self.stack.get(offset) && let Some(current_fp) = self.stack.get(offset-1) {
```

Comment:
Perhaps this could be a method

```rust
fn next_address(&self, fp: u64) -> Option<(u64, u64> {
  let offset = ((fp - self.stack_start - 2) / 8) as usize;
  if offset == 0 {
     return None;
  }
  let address = self.stack.get(offset)?;
  let current_fp = self.stack.get(offset - 1)?;

  Some(address, current_fp)
```

That means you can then just do `if let Some(address, current_fp) = self.next_address()`

Reply:
The problem with this is that you lose information (if offset or offset -1 are not in the stack).

Reply:
I'm not sure I follow? If either of `offset` or `offset - 1` some part of the if statement will be None and not executed. This function does the same thing by prematurely returning `None` via ?

Reply:
My bad. You are right, although I had to add a few *'s